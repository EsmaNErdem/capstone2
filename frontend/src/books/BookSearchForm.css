.SearchBox {
    margin: 20px;
    padding: 20px;
    border: 1px solid yellowgreen;
    border-radius: 8px;
    transition: background-color 0.3s ease;
  }
  
  .AdvancedSearch {
    background-color: #f8f8f8;
  }
  
  .SearchForm {
    display: flex;
    flex-wrap: wrap;
    gap: 10px;
  }
  
  .SearchForm input {
    flex-grow: 1;
    padding: 10px;
    border: 1px solid #ccc;
    border-radius: 4px;
    font-size: 16px;
  }

  .SearchInputContainer {
    position: relative;
    flex-grow: 1;
  }
  
  .SearchInputContainer input {
    width: 90%;
    padding: 10px;
    border: 1px solid #ccc;
    border-radius: 4px;
    font-size: 16px;
  }
  
  .SubmitButton-inside {
    position: absolute;
    top: 50%;
    right: 5%;
    transform: translateY(-50%);
    background-color: #f8f8f8;
    color: gray;
    padding: 10px 15px;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    transition: background-color 0.3s ease;
  }
  
  .ToggleButton {
    background-color: yellowgreen;
    color: white;
    padding: 10px 15px;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    transition: background-color 0.3s ease;
  }

  .ToggleButton-large {
    background-color: yellowgreen;
    color: white;
    padding: 10px 50px;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    transition: background-color 0.3s ease;
  }
  
  .ToggleButton:hover {
    background-color: #45a049;
  }
  
  .SubmitButton {
    background-color: #f00000;
    color: white;
    padding: 10px 15px;
    border: none;
    border-radius: 4px;
    cursor: pointer;
    transition: background-color 0.3s ease;
  }
  
  .SubmitButton:hover {
    background-color: rgb(174, 17, 17);
  }  