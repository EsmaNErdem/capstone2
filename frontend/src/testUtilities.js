import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
