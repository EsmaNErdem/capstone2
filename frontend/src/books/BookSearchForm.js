import React, { useState, useEffect, useRef } from "react";
import Prompt from "../utilities/Prompt";
import "./BookSearchForm.css"; 
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faFilter, faSearch } from '@fortawesome/free-solid-svg-icons';

/**
 *  Displays reusable search box component for searching book
 * 
 * - search value searches for value matching book title, authors or publisher.
 * - detail search with terms: { title, author, publisher, subject } 
 * - handleSubmit triggers API call in related extension
 * 
 * This component doesn't *do* the searching, but it renders the search
 * form and calls the `searchFor` function prop that runs in a parent to do the
 * searching.
 * 
 * - BookCard ==> BookSearchForm
 */
const BookSearchForm = ({searchFor}) => {
    console.debug("BookSearchForm", "searchFor=", typeof searchFor);

    const [formData, setFormData] = useState({})
    const [advancedSearch, setAdvancedSearch] = useState(false)
    const timeoutId = useRef();

    const sendSearchData = () => {
        let title
        let author
        let publisher
        let subject 

        if (formData.title) title = formData.title.trim()
        if (formData.author) author = formData.author.trim()
        if (formData.publisher) publisher = formData.publisher.trim()
        if (formData.subject) subject = formData.subject.trim()

        const searchData = {
            search:
                formData.search ||
                title ||
                author ||
                publisher ||
                subject,
            terms: {
                title,
                author,
                publisher,
                subject
            },
        };

        if(searchData.search) searchFor(searchData || undefined)
      }

    //  Handles form submission and sends searchData 
    const handleSubmit = e => {
        e.preventDefault();
        sendSearchData();
    }
    
    // Handles changes in form inputs and updates the component's state accordingly.
    // if it is checkbox input, it becomes "true" string for query parameter in API call
    // for other inputs, turn empty string into undefined 
    const handleChange = (prompt, value, e) => {
        setFormData(data => ({
            ...data,
            [prompt] : value || undefined
        }))

        if (!advancedSearch) { 
            if (timeoutId.current) {
                clearTimeout(timeoutId.current);
            }
        }  
    }

    // Clear the timeout when the component unmounts or when input changes
    useEffect(() => {
        if (!advancedSearch) {
            if (timeoutId.current) {
                clearTimeout(timeoutId.current);
            }
            
            timeoutId.current = setTimeout(() => {
                sendSearchData();
            }, 1200);

            return () => {
                if (timeoutId.current) {
                clearTimeout(timeoutId.current);
                timeoutId.current = null
                }
            };
        }    
    }, [formData]);

    // Toggle advancedSearch state
    const toggleAdvancedSearch = () => {
        setAdvancedSearch((prev) => !prev);
    };

    return (
        <div className={`SearchBox ${advancedSearch ? "AdvancedSearch" : ""}`}>
            <form className="SearchForm" onSubmit={handleSubmit}>
                <div className="SearchInputContainer">
                    <Prompt 
                        prompt={"search"} 
                        value={formData["search"]} 
                        handleChange={handleChange}
                    />
                    {!advancedSearch && (
                        <FontAwesomeIcon icon={faSearch} className="SubmitButton-inside" />
                    )}
                </div>
                {advancedSearch ? (
                    <>
                        {["title", "author", "publisher", "subject"].map(prompt => (
                            <Prompt key={prompt}
                                prompt={prompt}
                                value={formData[prompt]}
                                handleChange={handleChange}
                            />
                        ))}
                        <button
                            type="button"
                            className="ToggleButton"
                            onClick={toggleAdvancedSearch}
                            title="Toggle Advanced Search"
                            >
                            <FontAwesomeIcon icon={faFilter} />
                        </button>
                        <button 
                            data-testid="submit"
                            type="submit" 
                            className="SubmitButton" 
                            title="Submit Search"
                            >
                            <FontAwesomeIcon icon={faSearch} /> 
                        </button>
                    </>
                    ) : (
                    <>
                        <button
                            data-testid="filter-toggle"
                            type="button"
                            className="ToggleButton-large"
                            onClick={toggleAdvancedSearch}
                            title="Toggle Advanced Search"
                            >
                            <FontAwesomeIcon icon={faFilter} />
                        </button>
                    </>
                )}
            </form>
        </div>
    )
}

export default BookSearchForm; 