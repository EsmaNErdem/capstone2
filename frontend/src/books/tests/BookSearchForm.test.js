import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../../testUtilities";
import { render, fireEvent, act, waitFor } from '@testing-library/react';
import BookSearchForm from '../BookSearchForm';

const searchFor = jest.fn();
jest.useFakeTimers();

afterEach(() => {
  jest.clearAllTimers();
})

test('renders without crashing', async () => {  
    await act(async () => {
      render(
        <MemoryRouter>
          <UserProvider>
            <BookSearchForm />
          </UserProvider>
        </MemoryRouter>
      );
    });
});

test('it renders and matches with snapshot', async () => {  
    await act(async () => {
      ({ asFragment, getByTestId, queryByPlaceholderText } = render(
        <MemoryRouter>
          <UserProvider>
            <BookSearchForm />
          </UserProvider>
        </MemoryRouter>
      ));
    });

    expect(asFragment()).toMatchSnapshot();
    
    const filterToggle = getByTestId('filter-toggle');
    fireEvent.click(filterToggle);

    expect(queryByPlaceholderText('Book search')).toBeInTheDocument()

    // Wait for state to update
    await waitFor(() => {
      expect(asFragment()).toMatchSnapshot();
    });

});
  
test('search redirects to the correct route', async () => {
  await act(async () => {
    ({ getByPlaceholderText, getByTestId } = render(
      <MemoryRouter>
        <UserProvider>
          <BookSearchForm searchFor={searchFor}/>
        </UserProvider>
    </MemoryRouter>
  ));
  });

  const searchInput = getByPlaceholderText('Book search');
  fireEvent.change(searchInput, { target: { value: 'Harry' } });

  fireEvent.submit(searchInput);
 
  expect(searchFor).toHaveBeenCalled();
  expect(searchFor).toHaveBeenCalledWith({
    search: "Harry",
    terms: {
           author: undefined,
           publisher: undefined,
           subject: undefined,
           title: undefined,
         }
  });
});

test('toggle more search inputs', async () => {
  await act(async () => {
    ({ queryByPlaceholderText, getByTestId } = render(
      <MemoryRouter>
        <UserProvider>
          <BookSearchForm />
        </UserProvider>
    </MemoryRouter>
  ));
  });

  const filterToggle = getByTestId('filter-toggle');
  expect(queryByPlaceholderText('Book search')).toBeInTheDocument()
  expect(queryByPlaceholderText('Book author')).not.toBeInTheDocument()
  fireEvent.click(filterToggle);
  expect(queryByPlaceholderText('Book author')).toBeInTheDocument()
});

// Mock the setTimeout function to control the passage of time in the tests
test('debounces the search input', async () => {

  await act(async () => {
    ({ getByPlaceholderText } = render(
      <MemoryRouter>
        <UserProvider>
          <BookSearchForm searchFor={searchFor}/>
        </UserProvider>
    </MemoryRouter>
  ));
  });

  const searchInput = getByPlaceholderText('Book search');
  await act(async () => {
    fireEvent.change(searchInput, { target: { value: 'Harry' } });
  });
  
  expect(searchFor).not.toHaveBeenCalled();
  
  // Move time forward by 1200ms (the debounce time)
  act(() => {
    jest.advanceTimersByTime(1200);
  });
  
  expect(searchFor).toHaveBeenCalled();
  // await waitFor(() => {
  //   expect(searchFor).toHaveBeenCalledWith({
  //     search: 'Harry',
  //     terms: {
  //       author: undefined,
  //       publisher: undefined,
  //       subject: undefined,
  //       title: undefined,
  //     },
  //   });
  // });
});  