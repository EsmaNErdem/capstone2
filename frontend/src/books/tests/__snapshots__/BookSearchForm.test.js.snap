// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snapshot 1`] = `
<DocumentFragment>
  <div
    class="SearchBox "
  >
    <form
      class="SearchForm"
    >
      <div
        class="SearchInputContainer"
      >
        <input
          placeholder="Book search"
          value=""
        />
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-magnifying-glass SubmitButton-inside"
          data-icon="magnifying-glass"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
            fill="currentColor"
          />
        </svg>
      </div>
      <button
        class="ToggleButton-large"
        data-testid="filter-toggle"
        title="Toggle Advanced Search"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-filter "
          data-icon="filter"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3.9 54.9C10.5 40.9 24.5 32 40 32H472c15.5 0 29.5 8.9 36.1 22.9s4.6 30.5-5.2 42.5L320 320.9V448c0 12.1-6.8 23.2-17.7 28.6s-23.8 4.3-33.5-3l-64-48c-8.1-6-12.8-15.5-12.8-25.6V320.9L9 97.3C-.7 85.4-2.8 68.8 3.9 54.9z"
            fill="currentColor"
          />
        </svg>
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`it renders and matches with snapshot 2`] = `
<DocumentFragment>
  <div
    class="SearchBox AdvancedSearch"
  >
    <form
      class="SearchForm"
    >
      <div
        class="SearchInputContainer"
      >
        <input
          placeholder="Book search"
          value=""
        />
      </div>
      <input
        placeholder="Book title"
        value=""
      />
      <input
        placeholder="Book author"
        value=""
      />
      <input
        placeholder="Book publisher"
        value=""
      />
      <input
        placeholder="Book subject"
        value=""
      />
      <button
        class="ToggleButton"
        title="Toggle Advanced Search"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-filter "
          data-icon="filter"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3.9 54.9C10.5 40.9 24.5 32 40 32H472c15.5 0 29.5 8.9 36.1 22.9s4.6 30.5-5.2 42.5L320 320.9V448c0 12.1-6.8 23.2-17.7 28.6s-23.8 4.3-33.5-3l-64-48c-8.1-6-12.8-15.5-12.8-25.6V320.9L9 97.3C-.7 85.4-2.8 68.8 3.9 54.9z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        class="SubmitButton"
        data-testid="submit"
        title="Submit Search"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-magnifying-glass "
          data-icon="magnifying-glass"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
            fill="currentColor"
          />
        </svg>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
